### PR TITLE
feat: Add support of Google access token #307

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: src/main/java/com/epam/aidial/core/security/AccessTokenValidator.java [hunk 1]
```diff
@@ -1,59 +1,120 @@
...
                 }
             }));
         }
+        vertx.setPeriodic(0, USER_INFO_EXP_PERIOD_MS, event -> evictExpiredUserInfo());
+    }
+
+    private void evictExpiredUserInfo() {
+        long currentTime = System.currentTimeMillis();
+        for (Map.Entry<String, Future<UserInfoResult>> entry : userInfoCache.entrySet()) {
+            Future<UserInfoResult> future = entry.getValue();
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/AccessTokenValidator.java]: Nit:`UserInfoResult userInfo = entry.getValue().result()`

File: src/main/java/com/epam/aidial/core/security/IdentityProvider.java [hunk 2]
```diff
@@ -78,6 +91,17 @@ public IdentityProvider(JsonObject settings, Vertx vertx, Function<String, JwkPr
             jwkProvider = jwkProviderSupplier.apply(jwksUrl);
         }
 
+        try {
+            String userinfoEndpoint = settings.getString("userinfoEndpoint");
+            if (userinfoEndpoint != null) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: Nit: `userinfoUrl = userinfoEndpoint != null ? new URL() : null;`

File: src/main/java/com/epam/aidial/core/security/IdentityProvider.java [hunk 3]
```diff
@@ -78,6 +91,17 @@ public IdentityProvider(JsonObject settings, Vertx vertx, Function<String, JwkPr
             jwkProvider = jwkProviderSupplier.apply(jwksUrl);
         }
 
+        try {
+            String userinfoEndpoint = settings.getString("userinfoEndpoint");
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: please, update README.md to describe new IdP setting

File: src/main/java/com/epam/aidial/core/security/IdentityProvider.java [hunk 4]
```diff
@@ -228,7 +278,26 @@ private Map<String, List<String>> extractUserClaims(DecodedJWT decodedJwt) {
...
     }
 
-    Future<ExtractedClaims> extractClaims(DecodedJWT decodedJwt) {
+    @SuppressWarnings("unchecked")
+    private Map<String, List<String>> extractUserClaims(JsonObject userInfo) {
+        Map<String, List<String>> userClaims = new HashMap<>();
+        for (Map.Entry<String, Object> entry : userInfo.getMap().entrySet()) {
+            String claimName = entry.getKey();
+            Object claimValue = entry.getValue();
+            if (claimValue instanceof String) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: Nit: `if (claimValue instanceof String stringClaimValue)`

File: src/main/java/com/epam/aidial/core/security/IdentityProvider.java [hunk 5]
```diff
@@ -238,8 +307,37 @@ Future<ExtractedClaims> extractClaims(DecodedJWT decodedJwt) {
         return verifyJwt(decodedJwt).map(this::from);
     }
 
+    Future<ExtractedClaims> extractClaimsFromUserInfo(String accessToken) {
+        RequestOptions options = new RequestOptions()
+                .setAbsoluteURI(userinfoUrl)
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: .addHeader("Authorization", "Bearer " + accessToken)

Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: fixed

File: src/main/java/com/epam/aidial/core/security/IdentityProvider.java (hunk 4 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: fixed

File: src/main/java/com/epam/aidial/core/security/IdentityProvider.java (hunk 2 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: fixed

File: src/main/java/com/epam/aidial/core/security/AccessTokenValidator.java (hunk 1 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/security/AccessTokenValidator.java]: fixed

File: src/main/java/com/epam/aidial/core/security/IdentityProvider.java (hunk 3 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/security/IdentityProvider.java]: updated.

File: src/main/java/com/epam/aidial/core/security/AccessTokenValidator.java [hunk 6]
```diff
@@ -1,59 +1,154 @@
...
+            accessToken = Objects.requireNonNull(extractTokenFromHeader(authHeader), "Can't extract access token from header");
             if (providers.size() == 1) {
-                return providers.get(0).extractClaims(jwt);
-            }
-            for (IdentityProvider idp : providers) {
-                if (idp.match(jwt)) {
-                    return idp.extractClaims(jwt);
-                }
+                IdentityProvider provider =  providers.get(0);
+                return extractClaims(accessToken, provider);
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/AccessTokenValidator.java:69]: I don't think that handling single IdP will benefit a lot. for now it only produce extra code

**Review by Maxim-Gadalov — DISMISSED** [review]

**Comment by Maxim-Gadalov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-317.nightly-test.deltixhub.io
>E2E tests status: failed

File: src/main/java/com/epam/aidial/core/security/GetUserRoleFunctionFactory.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    private Future<List<String>> getGoogleWorkspaceGroups(String accessToken, Map<String, Object> claims) {
+        String userEmail = (String) claims.get("email");
+        RequestOptions options = new RequestOptions()
+                .setAbsoluteURI(GOOGLE_IAM_GROUPS_ENDPOINT.formatted(userEmail))
+                .setMethod(HttpMethod.GET);
+        Promise<List<String>> promise = Promise.promise();
+        client.request(options).onFailure(promise::fail).onSuccess(request -> {
+            request.putHeader(HttpHeaders.AUTHORIZATION, "Bearer " + accessToken);
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/GetUserRoleFunctionFactory.java]: you can move `putHeader` to options

Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/security/GetUserRoleFunctionFactory.java]: fixed

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-317.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-317.nightly-test.deltixhub.io
>E2E tests status: success

**Review by Maxim-Gadalov — APPROVED** [review]

**Pull request merged**